src/lib/Simplify.hs

```diff
@@ -76,13 +76,23 @@ simplifyDecls (Nest decl rest) = do
   return (substEnv <> substEnv')
 
 simplifyDecl :: Decl -> SimplifyM SubstEnv
+simplifyDecl (Let NoInlineLet (Bind (name:>_)) expr) = do
+  x <- simplifyStandalone $ Block Empty expr
+  emitTo name NoInlineLet (Atom x) $> mempty
 simplifyDecl (Let ann b expr) = do
   x <- simplifyExpr expr
   let name = binderNameHint b
   if isGlobalBinder b
     then emitTo name ann (Atom x) $> mempty
     else return $ b @> x
 
+simplifyStandalone :: Block -> SimplifyM Atom
+simplifyStandalone (Block Empty (Atom (LamVal b body))) = do
+  b' <- mapM substEmbedR b
+  buildLam b' PureArrow $ \x ->
+    extendR (b@>x) $ simplifyStandalone body
+simplifyStandalone block = simplifyBlock block
```

Comment:
Shouldn't marking a non-function atom as no-inline be an error? We don't actually take it into account when e.g. inlining tables.